astarte/device/device_grpc.py

```diff
@@ -582,25 +620,26 @@ def _decode_astarte_data_type_individual(astarte_data_type_individual: AstarteDa
     obj
         An object containig the decoded astarte_data_type_individual
     """
-    # All the individual_data options that contain arrays.
-    array_types = [
-        AstarteBooleanArray,
-        AstarteStringArray,
-        AstarteDoubleArray,
-        AstarteIntegerArray,
-        AstarteLongIntegerArray,
-        AstarteBinaryBlobArray,
-        AstarteDateTimeArray,
-    ]
-
-    individual_data_opt = astarte_data_type_individual.WhichOneof("individual_data")
-    individual_data = getattr(astarte_data_type_individual, individual_data_opt)
-    if individual_data_opt == "astarte_date_time":
-        individual_data = individual_data.ToDatetime(timezone.utc)
-    if any(isinstance(individual_data, array_type) for array_type in array_types):
-        individual_data = list(individual_data.values)
-    if individual_data_opt == "astarte_date_time_array":
-        individual_data = [e.ToDatetime(timezone.utc) for e in individual_data]
+
+    individual_data_opt = astarte_data_type_individual.WhichOneof("astarte_data")
+    proto_individual_data: ProtobufAstarteDataTypes = getattr(
+        astarte_data_type_individual, individual_data_opt
+    )
+
+    individual_data: AstarteDataTypes
+
+    if isinstance(proto_individual_data, Timestamp):
+        individual_data = proto_individual_data.ToDatetime(timezone.utc)
+    elif isinstance(proto_individual_data, typing.get_args(ProtobufAstarteDataVector)):
+        if isinstance(proto_individual_data, AstarteDateTimeArray):
+            individual_data = [e.ToDatetime(timezone.utc) for e in proto_individual_data.values]
+        else:
+            individual_data = list(proto_individual_data.values)
+    elif isinstance(proto_individual_data, typing.get_args(ProtobufAstarteDataScalar)):
+        individual_data = proto_individual_data
+    else:
+        logging.error("Unexpected individual type: %s", proto_individual_data)
```

Comment:
This function will most likely result in an exception since  `individual_data` is not defined when the function exits.
It would be better to raise a meaningful exception ourselves.